### ISSUE TITLE
pkg-config Windows

### ISSUE THREAD

**Issue by erakis** [mesonbuild/meson#4029]
Hi,

I'm trying for the first time the meson build system. Get it working on linux is easy but on Windows I'm a little bit stuck.

My project depends on `protobuf`.

1. Meson seems to use `pkg-config` to test the presence of dependency but where to find the latest pkg-config  for **Windows** ? I'm getting the error : **ERROR:  Pkg-config not found.**
2. Once I have installed the pkg-config, it will be a certainty that the `protobuf` library will not be available on the OS. So, which Windows compatible package manager should I install ? 
    * [vc-pkg](https://github.com/Microsoft/vcpkg) ? 
    * [Chocolatey](https://chocolatey.org/) ?
    * No one as no package manager on Windows can provide `*.pc` file ?

Best regards,

**Comment by purell** [comment]
Hi @jpakkane 

My project depends on (`protobuf 3`, `libsodium` and `zlib`). What do you advise me ?

What I want to achieve is to 
* Install dependencies
* Produce a Visual Studio solution
* Let the user open the solution and use it

**Comment by jpakkane** [comment]
You need to have some way of getting the dependencies. I don't use Windows all that much so I can't really say what the best way to do that is. Also note that the solution files produced by Meson are not portable between machines. Anyone who wants to use your project also needs to obtain all the dependencies and generate the solution files with Meson.

**Comment by purell** [comment]
The only way I found yet is 
1. Get source and compile library manually or use `vc-pkg` to get binary
2. I have to inform `pkg-config` where to find the library. Usually this is done by indicating the path (`PKG_CONFIG_PATH`) where are the `*.pc` file. But those do not even exists. Worst, they must contains the path where the library are, so I can ship `*.pc` file with hard-coded path.

This is a mess on Windows.

I think I will create a Visual Studio 2015 solution for Windows and a makefile for unix based os.

**Comment by barcharcraz** [comment]
My vote is
1) get pkgconf's meson build working on windows
2) allow embedding of pkgconf as a subproject, and make dependencies/base.py find it
3) use pkgconf's built in windows personality support with whatever .pc files are around.

3 works. 2 needs some changes in meson, and 1 is in progress (I've opened upstream issues).

**Comment by purell** [comment]
What about supporting [vc-pkg](https://github.com/Microsoft/vcpkg) to querying installed libraries ? pkg-config does not work as well on windows as linux.

From what I've test yet, Meson work perfectly on Linux but on Windows I will unfortunately use cmake with vc-pkg. As they are better coupled. But I hate cmake !

For Meson + Windows a package manager like vc-pkg is essential. It's too much painful having to manually compile each third party dependency, create or link `pk-config` file, etc....

**Comment by barcharcraz** [comment]
well vcpkg doesn't provide any functionality for finding libraries, the integrations it has happen via either setting all the link options globally (for msbuild) or adding WhateverConfig.cmake files to the cmake paths for cmake.

Neither are really viable for meson. However packages can output pc files on windows and they work just as well as on linux (pkgconf has some special handling of paths and whatnot). Relocation works as well. 

There has been some discussion in the cmake mailing lists around how to better support this sort of thing. Basically CmakeConfig.cmake files are quite coupled to cmake itself, and almost impossible to use from other tools, there's work going on to make generation of pkgconfg files easier from cmake and to encourage that.

**Comment by nirbheek** [comment]
You don't need pkg-config to use Meson, and I don't think it makes sense to *build* pkgconf to get it. You either have dependencies already available on your Windows system via pkg-config, but if you don't, then there are better ways of getting them.

One way would be with [nuget](https://cgit.freedesktop.org/gstreamer/gstreamer-sharp/tree/meson.build#n130) or vcpkg. Another would be to host and ship manually-created binary wraps for the dependencies that are painful to build (such as libsodium). Currently, the way to do this is to [download the binaries are configure time](https://cgit.freedesktop.org/gstreamer/gst-build/tree/subprojects/win-flex-bison-binaries/meson.build) and set up a `declare_dependency()` pointing to it.

There are already source wraps on wrapdb for zlib and protobuf. We are planning on adding better integration in Meson for binary wraps for Windows so you don't have to do the above work manually.

**Comment by purell** [comment]
@nirbheek Thank you so much for your explanation.

> You don't need pkg-config to use Meson
I was reading documentation about [declare_dependency](http://mesonbuild.com/Reference-manual.html#declare_dependency). Does it means I could use binary got from vc-pkg, a configure them manually ? The only drawback with this solution is that I will have to have a different configuration between linux and windows like.

```bash
if meson.get_compiler('cpp').get_id() != 'msvc'
    declare_dependency(
        include_directories: '<vc-pkg-installation-path>\installed\x86-windows`, 
        link_with: 'zlib libsodium protobuf`, 
    )
else
    zlib = dependency('zlib', version: '>=1.2')
    protobuf = dependency('protobuf', version: '>=3.5')
    sodium = dependency('libsodium', version: '>=1.0')
endif

executable('myapp',
           dependencies: [zlib, protobuf, sodium])
```

The way I try with Meson is : 
1. Install [pkg-config lite](https://sourceforge.net/projects/pkgconfiglite/) for Windows
2. Get [vc-pkg](https://github.com/Microsoft/vcpkg) and install dependencies
    ```bash
    vcpkg install zlib protobuf libsodium
    ```
3. Configure pkg-config using environment variable `PKG_CONFIG_PATH`  with `<vc-pkg-installation-path>\installed\x86-windows`
    ```bash
    set PKG_CONFIG_PATH=<vc-pkg-installation-path>\installed\x86-windows;%PKG_CONFIG_PATH%
    ``` 
4. Call meson build.

But I was always getting the error **ERROR: Pkg-config not found**. So I tried listing pkg installed package using
```bash
pkg-config --list-all
```
The only library listed was `protobuf`, no `zlib` or `libsodium`. The problem is that vc-pkg does not always produce `*.pc` file needed by `pkg-config`.

Looking at the original [cmake](https://github.com/madler/zlib/blob/v1.2.11/CMakeLists.txt#L225) of [zlib](https://github.com/madler/zlib/), the variable `SKIP_INSTALL_FILES` should NOT be set to get the `*.pc` file. But on the [vc-pkg porting](https://github.com/Microsoft/vcpkg/blob/master/ports/zlib/portfile.cmake#L23) they set this variable to `ON`.

**Comment by purell** [comment]
> Are you sure pkg-config is available in PATH? What does where pkg-config return?
```bash
> where pkg-config
C:\pkg-config-lite-0.28-1\bin\pkg-config.exe
```

But there only protobuf that output a `*.pc` file
```bash
D:\Data\vcpkg\installed\x86-windows>dir /s *.pc

Directory of D:\Data\vcpkg\installed\x86-windows\debug\lib\pkgconfig
2018/08/15  01:40 PM               425 protobuf-lite.pc
2018/08/15  01:40 PM               425 protobuf.pc
               2 File(s)            850 bytes

Directory of D:\Data\vcpkg\installed\x86-windows\lib\pkgconfig
2018/08/15  01:40 PM               401 protobuf-lite.pc
2018/08/15  01:40 PM               401 protobuf.pc
```
So, this is normal that `pkg-config` only list the `protobuf` as installed library
```bash
> echo %PKG_CONFIG_PATH%
D:\Data\vcpkg\packages\protobuf_x86-windows\lib\pkgconfig

> pkg-config --list-all
protobuf-lite Protocol Buffers - Google's Data Interchange Format
protobuf      Protocol Buffers - Google's Data Interchange Format
```

But for `zlib` and `libsodium` the vc-pkg configuration does not instruct to output the `*.pc` file 😞 

> sodium = cc.find_library('sodium')

I'm curious, what is `cc` and what location `find_library` will look ?

**Comment by mqudsi** [comment]
Writing custom .pc files is pretty straight forward, you can set `PKG_CONFIG_PATH` to a local directory and include the pc files in your version control for Windows.

**Comment by purell** [comment]
> Writing custom .pc files is pretty straight forward, you can set PKG_CONFIG_PATH to a local directory and include the pc files in your version control for Windows.

Interesting but if I use the package-manager `vk-pkg` to get binaries, they can be put anywhere. From what I've understood, the `pkg-config` file should contains the path where the binaries are. I can't push a `pkg-config` file with an hard-coded path to my version control repo.

**Comment by purell** [comment]
> This will use the compiler to find the library in the compiler's default search paths. See: 

@nirbheek Seriously on Windows this is a nightmare to use meson. If we use `vcpkg`, the binaries can be installed anywhere on the user's hard disk. 

Also, some package does not comes with `pkg-config` so meson can't find them automatically. As exemple, `protobuf` is coming with, but `zlib` and `libsodium` no.

I could create my own `pkg-config` file for those package and save them in my repo, but this will still make no sens as I should put an hard-coded path inside. Remember, `vc-pkg` can installed binaries anywhere.

Now trying your solution to make a custom search....
```bash
meson.get_compiler('cpp').find_library('zlib', dirs : ['...', '...'])
```

Here is the listening of my `vc-pkg` folder
```bash
D:\Data\vcpkg\installed>tree
D:.
├───vcpkg
│   ├───info
│   └───updates
├───x64-windows
│   ├───bin
│   ├───debug
│   │   ├───bin
│   │   └───lib
│   │       └───pkgconfig
│   ├───include
│   │   ├───google
│   │   │   └───protobuf
│   │   │       ├───compiler
│   │   │       │   ├───cpp
│   │   │       │   ├───csharp
│   │   │       │   ├───java
│   │   │       │   ├───js
│   │   │       │   ├───objectivec
│   │   │       │   ├───php
│   │   │       │   ├───python
│   │   │       │   └───ruby
│   │   │       ├───io
│   │   │       ├───stubs
│   │   │       └───util
│   │   └───sodium
│   ├───lib
│   │   └───pkgconfig
│   ├───share
│   │   ├───libsodium
│   │   ├───protobuf
│   │   └───zlib
│   └───tools
│       └───protobuf
└───x86-windows
    ├───bin
    ├───debug
    │   ├───bin
    │   └───lib
    │       └───pkgconfig
    ├───include
    │   ├───google
    │   │   └───protobuf
    │   │       ├───compiler
    │   │       │   ├───cpp
    │   │       │   ├───csharp
    │   │       │   ├───java
    │   │       │   ├───js
    │   │       │   ├───objectivec
    │   │       │   ├───php
    │   │       │   ├───python
    │   │       │   └───ruby
    │   │       ├───io
    │   │       ├───stubs
    │   │       └───util
    │   └───sodium
    ├───lib
    │   └───pkgconfig
    ├───share
    │   ├───libsodium
    │   ├───protobuf
    │   └───zlib
    └───tools
        └───protobuf
```

I can't put hard-coded path here too, this make no sens. More over, I can't use [environment variable](https://github.com/mesonbuild/meson/issues/9) easily too or passing argument to meson and use it in the `dirs` parameters of the function `find_library`.

In conclusion, I'm stuck and need to go back to cmake :(

**Comment by purell** [comment]
Other thing I tried is to add a path where the compiler should look.

```meson
add_project_link_arguments(['D:\Data\vcpkg\installed\x86-windows', 'D:\Data\vcpkg\installed\x86-windows\lib', 'D:\Data\vcpkg\installed\x86-windows\bin'], language : 'cpp')
```

But it still doesn't work. I'm getting
```
-----
Running test binary command: D:\Data\Projects\xxx\build-windows\meson-private\sanitycheckcpp.exe
Native C++ compiler: cl (msvc 19.00.24215.1)
Build machine cpu family: x86
Build machine cpu: x86
Running compile:
Working directory:  C:\Users\xxx\AppData\Local\Temp\tmpx_phz782
Command line:  cl C:\Users\xxx\AppData\Local\Temp\tmpx_phz782\testfile.cpp /nologo /showIncludes /FeC:\Users\xxx\AppData\Local\Temp\tmpx_phz782\output.exe /MDd /Od zlib.lib 
```

Why there is no 
```
/LIBPATH:D:\Data\vcpkg\installed\x86-windows; D:\Data\vcpkg\installed\x86-windows\lib;D:\Data\vcpkg\installed\x86-windows\bin
``` 
add to the linker ? If this was the case, the compiler should find the library.

**Comment by barcharcraz** [comment]
I bet we could generate pkg config files from the cmake stuff vcpkg
provides, as long as we limit ourselves to just vcpkg and not cmake in
general. It might also be good to prod vcpkg upstream to start providing
pkgconfig files.

**Comment by purell** [comment]
@barcharcraz I hijack an [issue on vc-pkg](https://github.com/Microsoft/vcpkg/issues/2662#issuecomment-417000587), hoping they recognizes the need to produce a pkg-config file for meson.

I took a look to the [zlib library](https://github.com/madler/zlib/blob/master/CMakeLists.txt#L225) 
```cmake
if(NOT SKIP_INSTALL_FILES AND NOT SKIP_INSTALL_ALL )
    install(FILES ${ZLIB_PC} DESTINATION "${INSTALL_PKGCONFIG_DIR}")
endif()
```

Now, looking to the [zlib porting](https://github.com/Microsoft/vcpkg/blob/master/ports/zlib/portfile.cmake#L23) of vc-pkg

```cmake
vcpkg_configure_cmake(
    SOURCE_PATH ${SOURCE_PATH}
    PREFER_NINJA
    OPTIONS
        -DSKIP_INSTALL_FILES=ON
        -DSKIP_BUILD_EXAMPLES=ON
    OPTIONS_DEBUG
        -DSKIP_INSTALL_HEADERS=ON
)
```
Unfortunately options are hardcoded and no pc file can be produced.

**Comment by ras0219-msft** [comment]
@purell We'd (vcpkg) be happy to welcome PRs that improve the pkg-config files! Ideally, we would have a `vcpkg_fixup_pkgconfig()` macro similar to cmake that would perform any needed modifications (such as making the files relocatable, moving them to a single consistent location, etc).

If we want to make .pc files consistently high quality, then we would also want to have post build checks (currently, we check things like CRT version, file layout, and ensuring the library built static/dynamic libraries based on what the user requested). We may also want post-install usage messages like we have for CMake:

```
The package zlib is compatible with built-in CMake targets:

    find_package(ZLIB REQUIRED)
    target_link_libraries(main PRIVATE ZLIB::ZLIB)
```

Additionally, there are a few "overarching" questions I have about how meson would consume these files -- such as how should debug vs release work? I know nothing about meson but I'm pretty familiar with CMake; if meson acts like a CMake "multi-config" generator, then there may be complex work to do regarding merging the two flavors. If Meson is always a single configuration, then having two separate .pc files is probably better and we'd need to store them in two separate locations (passing the right location in to Meson based on the build configuration).

(https://github.com/Microsoft/vcpkg/pull/4155 may be relevant)

**Comment by galexite** [comment]
@ras0219-msft Could vcpkg export its own alternative configuration (i.e. as a JSON file) or provide a command similar to `pkg-config`, consumable by third-party build systems (such as Meson), thus allowing greater access to the features vcpkg provides?

I suggested this sometime back in an issue mentioned by @purell: Microsoft/vcpkg#2662

**Comment by purell** [comment]
The good news is that I finally found a way to get cpp compiler of meson (VS) able to find the library with no pkg-config file. The trick is to set the [LIB ](https://docs.microsoft.com/en-us/previous-versions/visualstudio/visual-studio-2010/6y6t9esh%28v%3dvs.100%29) environment variable before calling meson.

In my case 
```
set LIB=%LIB%;d:\vcpkg\installed\x86-windows\\lib;
meson ..
```

This way, the compiler `cl.exe` will try to find the library in installed folder of vc-pkg. But I have to make an exception for Windows vs Unix.

```meson
if build_machine.system() == 'windows'
  cpp = meson.get_compiler('cpp')
  zlib = cpp.find_library('zlib')
  sodium = cpp.find_library('libsodium')
else
  zlib = dependency('zlib', version: '>=1.2')
  sodium = dependency('libsodium', version: '>=1.0')
endif
protobuf = dependency('protobuf', version: '>=3.0')
```

This is not a perfect solution, but I hope this will help someone else. 

Obviously, it would be much simpler if `vcpkg` would provides the pkg-file. I'll see if I can find time to do a PR, but seriously I'm pretty new with cmake.

**Comment by purell** [comment]
I shouted victory too quickly.

Now meson is happy but once I start the compilation of the Visual Studio solution generated, I'm getting a link error. VS don't produce a lib file for my dll. Thus, my executable cannot be link with my dll.

With cmake I used the option [CMAKE_WINDOWS_EXPORT_ALL_SYMBOLS](https://cmake.org/cmake/help/v3.4/prop_tgt/WINDOWS_EXPORT_ALL_SYMBOLS.html#prop_tgt:WINDOWS_EXPORT_ALL_SYMBOLS) that automatically create a module definition (.def) file with all global symbols. 

I would like to do the same using `dumpbin` but meson lack of `prelink/postlink` task. **Why ?** Most of build system propose that.

Else i could
*. use the old way `__declspec(dllexport)` but 😞 
*. maintain a def file, I find it painful.

**Comment by barcharcraz** [comment]
that cmake feature internally makes a def file. I’d encourage the use of
dllexport and the equivalents in gcc.

**Comment by fungos** [comment]
I'm having trouble with this also and I believe the right fix is that Meson support an new alternate way of finding libraries. The LIB trick from @purell is a hint on this direction, but in his case he needed to adapt a meson.build file. Maybe this can be a fallback if either pkg_config and the cmake alternatives do not work.

For context, my issue is getting GTK to build on Windows, with Meson is basically impossible on windows to find all dependencies.

**Comment by agurtovoy** [comment]
@fungos I highly recommend looking into using Meson together with Conan (https://conan.io/). Conan truly solves the dependency problem across all platforms (including installing build prerequisites like pkg-config), and Meson shines as a high-level build system; they really work together quite nicely.

**Comment by HE-Wenjian** [comment]
> 
> 
> @fungos I highly recommend looking into using Meson together with Conan (https://conan.io/). Conan truly solves the dependency problem across all platforms (including installing build prerequisites like pkg-config), and Meson shines as a high-level build system; they really work together quite nicely.

I also use meson+Conan. How do you make Conan work together with meson? I currently use the code below to first search a dependency in system, then if unavailable, ask Conan to download given a `XXX.txt` conanfile and generate a `FindXXX.cmake`. This shall provide flexibility and work both on Linux and Win. However, if I change the `XXX.txt` conanfile, e.g. change the package version, meson reconfiguration fails to respect the newest `FindXXX.cmake` due to its dependency caching mechanism. Do you know how to solve that?

```meson
search_sys_then_conan = {
    'fmt':{'version':['>=5','<6']}
    # ... the dependency list goes on
}
deps = []
foreach pkg_name, config : search_sys_then_conan 
    pkg = dependency(pkg_name, required: false, kwargs: config )
    if not pkg.found()
        conan = find_program('conan', required: true)
        conanfile=files('./conanfiles/'+pkg_name+'.txt')
        conan_gen = configure_file(
            output : 'Find'+pkg_name+'.cmake',
            input : conanfile,
            command : [conan,'install',conanfile,'-if',meson.current_build_dir()])
        pkg = dependency(pkg_name, kwargs: config, 
            method: 'cmake', cmake_module_path : meson.current_build_dir())
    endif
    deps += pkg
endforeach

executable('exe_need_deps',  ['main.cpp'],
    dependencies: deps
)
```

```tree
project/
├── conanfiles/
│   └── fmt.txt
├── main.cpp
└── meson.build
```
`conanfiles/fmt.txt`:
```conanfile
[requires]
fmt/5.3.0
# Comment: If I change the line above, meson can trigger a reconfiguration, but will fail to adapt it.

[generators]
cmake_find_package

```

**Comment by agurtovoy** [comment]
@HE-Wenjian My setup is more traditional, I just let `conan` drive the build as per https://docs.conan.io/en/latest/reference/build_helpers/meson.html

**Comment by eli-schwartz** [comment]
I think this issue is no longer as relevant as it used to be -- if it's still relevant at all. Additionally, there seems to have been some misinformation going on here. I'll try to clarify the current state of affairs for how this works.

The most important thing to consider is that Meson's dependency lookups are fully decoupled from the tool used to look them up. The preferred mechanism is typically pkg-config, but:
- even long before this issue was opened, that could be overridden in Meson's source code and have Meson expose carefully curated custom dependency types for:
  - hand-rolled find_library and has_header checks
  - foo-config shell scripts that print their info using `foo-config --cflags --libs`
- in Meson 0.49, not long after this PR was opened in fact, Meson gained the ability (see commit a0175ecb1412ed1c824e5eaf1629be9ed95f0b78) to have `dependency('foo')` look up `Findfoo.cmake` or `foo-config.cmake` if that dependency is not found using pkg-config. Since cmake's format is fully internal and requires a full cmake language parser as well as various internal cmake environment state, this means building a dummy cmake project that looks up the dependency, then dumps debug traces in a documented json format, which is then parsed by Meson in order to read the necessary compiler flags and located libraries. Painful, requires installing both meson and cmake, but also works
- ideally, the dependency name is the same for pkg-config and for cmake; this is not actually always true, particularly when it comes to the *case* of the dependency name, e.g. `foo.pc` vs. `Foo-config.cmake` (I do not know why cmake packages are so fond of having capital names, but there you have it), and this may or may not work correctly on systems with case-***in***sensitive filesystems, but definitely won't work on systems with case-sensitive filesystems. Since Meson 0.6.0, you can check both like this: `dependency('foo', 'Foo')`

Underlying all this is the assumption that pkg-config "doesn't work" on Windows. It's true that many projects don't install .pc files, most particularly if they use cmake (which has a NIH bias against pkg-config because they didn't develop pkg-config personally), and also if... the package manager strips .pc files out because they think it is useless on Windows.

People especially like suggesting that pkg-config is useless on Windows because relocatable, which pkg-config doesn't support. A little-known fact about pkg-config is that it *does* support relocatable installs. Instead of embedding absolute paths in a .pc file, such as `prefix=C:/Software Bundles/foobar/` which then breaks if you move the whole directory elsewhere, you can specify paths in a .pc file like this: `prefix=${pcfiledir}/../../` -- it is then calculated based on the current location of the .pc file itself. This is the same thing that cmake calls `${CMAKE_CURRENT_LIST_DIR}`.

Meson recently gained support for generating relocatable .pc files itself in #10277 (not yet released). It's also possible to patch that out in conan/vcpkg/etc. and some projects also patch that out in configure.ac for autotools `*.pc.in` file templates but that does unfortunately require projects to explicitly support it where the build system doesn't have a native pkgconfig generator that supports this already.

Of course, you do need pkg-config built for Windows, but pkgconf works pretty well there these days, as I understand things. And your C++ package manager might already install it for you, if that package manager is named Conan. :)

**Comment by xclaesse** [comment]
> Of course, you do need pkg-config built for Windows

Meson might gain its own python implementation as well, pending review: https://github.com/mesonbuild/meson/pull/9021.
